### PR TITLE
#12805: Migrate moreh_sum_backward operation from tt_eager to ttnn

### PR DESCRIPTION
**Pull request by ngohoang34**
### Ticket
Link to Github Issue: [#12805](https://github.com/tenstorrent/tt-metal/issues/12805)

### Problem description
`moreh_sum_backward` was deprecated alongside `tt_dnn`. This PR ports it to `ttnn` using its operation format.

### What's changed
- Removed `moreh_sum_backward`'s tt_dnn implementations
- Move device code to `ttnn`
- Create new wrapper code for `ttnn` with new modules

### Checklist
- [x] Post commit CI passes:
+ [All post-commit tests #16044](https://github.com/tenstorrent/tt-metal/actions/runs/10938570502)
+ [All post-commit tests #16784](https://github.com/tenstorrent/tt-metal/actions/runs/11136918744)
- [x] Blackhole Post commit (if applicable) - NA
- [x] Model regression CI testing passes (if applicable) - NA
- [x] New/Existing tests provide coverage for changes 227/227

### CODE REVIEW

**Review by mrshaw01 — APPROVED** [review]
Revise phase done! Please review!
Detail: [GitHub Link](https://github.com/tenstorrent/tt-metal/compare/991e3101ba3b7aecda831349d069b3f7a2e929d8..b3bdc9968a57ae0bae7fc15e9288e1f1df58f600)

File: ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/device/moreh_sum_backward_device_operation.hpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#include "ttnn/decorators.hpp"
+#include "ttnn/device_operation.hpp"
+#include "ttnn/operations/core/compute_kernel/compute_kernel_config.hpp"
+
+namespace ttnn::operations::moreh::moreh_sum_backward {
+struct MorehSumBackwardOperation {
+    struct operation_attributes_t {
+        const std::vector<int64_t> dims;
+        const bool keepdim;
+        const MemoryConfig input_grad_mem_config;
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/device/moreh_sum_backward_device_operation.hpp:14]:
```suggestion
        const MemoryConfig input_grad_memory_config;
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/device/moreh_sum_backward_device_operation.hpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        const std::vector<int64_t>& dims,
+        bool keepdim,
+        const std::optional<Tensor>& input_grad,
+        const MemoryConfig& input_grad_mem_config,
+        const DeviceComputeKernelConfig& compute_kernel_config);
+};
+}  // namespace ttnn::operations::moreh::moreh_sum_backward
+
+namespace ttnn::prim {
+constexpr auto moreh_sum_backward = ttnn::register_operation_with_auto_launch_op<
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/device/moreh_sum_backward_device_operation.hpp:69]:
```suggestion
constexpr auto moreh_sum_backward = ttnn::register_operation<
```
Primitive ops should use register_operation

File: ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/moreh_sum_backward.cpp [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    uint32_t rank = input.has_value() ? input->get_shape().value.rank() : input_grad->get_shape().value.rank();
+    std::vector<int64_t> dims = tt::operations::primary::get_dim(dim, rank);
+    std::sort(dims.begin(), dims.end());
+    return ttnn::prim::moreh_sum_backward(
+        output_grad,
+        input,
+        dims,
+        keepdim,
+        input_grad,
+        input_grad_mem_config.value_or(operation::DEFAULT_OUTPUT_MEMORY_CONFIG),
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/moreh_sum_backward.cpp:29]:
```suggestion
        input_grad_mem_config.value_or(output_grad.memory_config()),
```
Using the default memory config can cause incorrect behavior where DRAM output is produced when L1 input is provided. In such cases, the memory config of the input should be used.

File: ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/device/moreh_sum_backward_device_operation.hpp [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    static void validate_on_program_cache_hit(const operation_attributes_t&, const tensor_args_t&);
+    static shape_return_value_t compute_output_shapes(const operation_attributes_t&, const tensor_args_t&);
+    static tensor_return_value_t create_output_tensors(const operation_attributes_t&, const tensor_args_t&);
+    static std::tuple<operation_attributes_t, tensor_args_t> invoke(
+        const Tensor& output_grad,
+        const std::optional<Tensor>& input,
+        const std::vector<int64_t>& dims,
+        bool keepdim,
+        const std::optional<Tensor>& input_grad,
+        const MemoryConfig& input_grad_mem_config,
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/device/moreh_sum_backward_device_operation.hpp:63]:
```suggestion
        const std::optional<MemoryConfig>& input_grad_memory_config,
```
memory_config should be optional until invoke.

File: ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/device/moreh_sum_backward_device_operation.hpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+#include "ttnn/device_operation.hpp"
+#include "ttnn/operations/core/compute_kernel/compute_kernel_config.hpp"
+
+namespace ttnn::operations::moreh::moreh_sum_backward {
+struct MorehSumBackwardOperation {
+    struct operation_attributes_t {
+        const std::vector<int64_t> dims;
+        const bool keepdim;
+        const MemoryConfig input_grad_memory_config;
+        const DeviceComputeKernelConfig compute_kernel_config;
```
Comment (hschoi4448) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/device/moreh_sum_backward_device_operation.hpp:15]: I'm waiting for a response from the TT side on this part. Everything else looks good.

Comment (hschoi4448) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/device/moreh_sum_backward_device_operation.hpp:15]: TT mentioned that optional depends on the use-case, so it seems it doesn't need to be optional.

**Review by hschoi4448 — APPROVED** [review]

**Review by razorback3 — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_device_operation.cpp [hunk 6]
```diff
@@ -136,12 +136,10 @@ MorehSumOperation::tensor_return_value_t MorehSumOperation::create_output_tensor
 std::tuple<MorehSumOperation::operation_attributes_t, MorehSumOperation::tensor_args_t> MorehSumOperation::invoke(
     const Tensor& input,
     const int64_t dim,
-    const bool keep_batch_dim,
+    const bool keepdim,
     const std::optional<Tensor>& output,
     const std::optional<MemoryConfig>& output_mem_config,
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_device_operation.cpp:141]: Should be `memory_config`

Comment (mrshaw01) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_device_operation.cpp:141]: Check other files for that.

File: ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/moreh_sum_backward_pybind.cpp [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+#include "moreh_sum_backward.hpp"
+#include "pybind11/decorators.hpp"
+
+namespace ttnn::operations::moreh::moreh_sum_backward {
+void bind_moreh_sum_backward_operation(py::module& module) {
+    bind_registered_operation(
+        module,
+        ttnn::moreh_sum_backward,
+        "Moreh moreh_sum_backward Operation",
```
Comment (mrshaw01) [inline ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/moreh_sum_backward_pybind.cpp:15]: Moreh Sum Backward Operation

**Review by mrshaw01 — APPROVED** [review]
LGTM. Please check the minors.

File: ttnn/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_device_operation.cpp (hunk 6 above)
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sum/device/moreh_sum_device_operation.cpp:141]: out of scope

File: ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/moreh_sum_backward_pybind.cpp (hunk 7 above)
Comment (ngohoang34) [reply ttnn/cpp/ttnn/operations/moreh/moreh_sum_backward/moreh_sum_backward_pybind.cpp]: resolved

**Review by ayerofieiev-tt — APPROVED** [review]

**Pull request merged**